### PR TITLE
[record]: Add `use=` field

### PR DESCRIPTION
**Pull request by erbesharat**
**What this PR does / why we need it**:
Adds the `use=` field to record to provide a feature like how CNAME records work. The `use=` field can be used to redirect a TXT record request to another upstream zone. Basically use a TXT record that has been specified for another zone address.

**Special notes for your reviewer**:
~~Still trying to figure out how to fix the bugs on the `path` type, so not ready for a full review.~~

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by erbesharat** [comment]
Found an edge-case on predefined regex records. Let's not merge this one till it gets fixed.

File: e2e/main.go [hunk 1]
```diff
@@ -177,14 +177,15 @@ func (d *dockerManager) StartContainers() error {
 }
 
 func (d *dockerManager) StopContainers() error {
-	if strings.Contains(d.dir, "dockerv2") {
-		_, err := exec.Command("docker",
+	if strings.Contains(d.dir, "path") {
```
Comment (stp-ip) [inline e2e/main.go]: Why the specific check here?

File: txtdirect.go [hunk 2]
```diff
@@ -173,12 +180,31 @@ func Redirect(w http.ResponseWriter, r *http.Request, c Config) error {
...
 
 	rec, err := getRecord(host, c, w, r)
-	r = rec.addToContext(r)
 	if err != nil {
 		fallback(w, r, "global", http.StatusFound, c)
 		return nil
 	}
 
+	// Add the upstream zone address from the use= fields to the request context
+	// and replace the source record with the upstream record
```
Comment (stp-ip) [inline txtdirect.go]: We don't replace the source record with the upstream record anymore or?

File: e2e/host/zonefile [hunk 3]
```diff
@@ -10,24 +10,29 @@
...
+_redirect.referer.host.host.example.com.         IN TXT "to=https://referer-redirect.host.host.example.com;type=host;ref=true"
 
-fallback.host.host.example.com.              IN A 172.20.10.2
-_redirect.fallback.host.host.example.com.    IN TXT "v=txtv0;to=https://{label30};type=host;code=302" 
+fallback.host.host.example.com.                  IN A 172.20.10.2
+_redirect.fallback.host.host.example.com.        IN TXT "v=txtv0;to=https://{label30};type=host;code=302" 
+
+sourcerecord.host.host.example.com.              IN A 172.20.10.2
+upstreamrecord.host.host.example.com.            IN A 172.20.10.2
+_redirect.sourcerecord.host.host.example.com.    IN TXT "v=txtv0;to=https://wrongredirect.host.host.example.com;type=host;code=302;use=_redirect.upstreamrecord.host.host.example.com"
```
Comment (stp-ip) [inline e2e/host/zonefile]: Interesting edge case that `to=` and `use=` are used at the same time. So `use=` takes precedent. Seems valid.

Comment (stp-ip) [reply e2e/host/zonefile]: Generally I feel like we should separately test this.

File: e2e/path/main.go [hunk 4]
```diff
@@ -74,6 +74,22 @@ var tests = []test{
...
+	{
+		name: "Redirect using the use= field and upstream record",
+		args: data{
+			host: "sourcerecord.path.path.example.com",
+			path: "/testing",
+		},
+		expected: "https://upstream.path.path.example.com",
+	},
+	{
+		name: "Redirect to upstream record using multiple use= fields",
```
Comment (stp-ip) [inline e2e/path/main.go]: Might be good to mention testing a non working upstream record and fallthrough to the next use= field. The description doesn't seem to explain the whole test well in my view.

File: e2e/path/main.go [hunk 5]
```diff
@@ -74,6 +74,22 @@ var tests = []test{
...
+		args: data{
+			host: "srcrecord.path.path.example.com",
+			path: "/testing",
+		},
+		expected: "https://upstream.path.path.example.com",
+	},
```
Comment (stp-ip) [inline e2e/path/main.go]: Adding an additional test, that uses `use=`, but that does go into fallback. Aka there is no valid upstream working.
That seems to be a good additional failure case.

File: e2e/main.go (hunk 1 above)
Comment (erbesharat) [reply e2e/main.go]: Oops. Leftover from debugging. Will remove it :)

File: txtdirect.go (hunk 2 above)
Comment (erbesharat) [reply txtdirect.go]: Removed the second line. Leftover from the previous implementation.

File: e2e/main.go (hunk 1 above)
Comment (erbesharat) [reply e2e/main.go]: Debugging Leftover. :sweat_smile:

File: record.go [hunk 6]
```diff
@@ -193,17 +193,20 @@ func ParseRecord(str string, w http.ResponseWriter, req *http.Request, c Config)
 		r.Code = http.StatusFound
 	}
 
-	if r.Type == "" {
-		r.Type = "host"
-	}
+	// Only apply rules and default to records that doesn't point to a upstream record
```
Comment (stp-ip) [inline record.go]: Good catch

Comment (stp-ip) [reply record.go]: Probably could be refactored into a defaulting function, but not sure if that makes it simpler.

**Comment by stp-ip** [comment]
Godocit failed. We should rename Godocit step from "build" to "Godocit" so it's easier to see in the PRs.

also not sure, why it failed so. Any ideas why the nil pointer dereference?
